### PR TITLE
feat(nargo): Support custom entry points specified in TOML

### LINKED ISSUE #1040: Allow configuring the source directory
### Problem

I mentioned in https://github.com/noir-lang/noir/issues/1009#issuecomment-1483926031 that foundry projects tend to use the `src` directory for contracts. This opens up an issue for if you want to use Noir within a foundry project, by default these will write noir and solidity code to the same directory

### Proposed solution

We could allow users to remap directories to custom paths by setting them in `Nargo.toml` similarly to how foundry does in `foundry.toml`.

https://book.getfoundry.sh/reference/config/project

### Alternatives considered

There are alternatives for the meantime:
- Perform this remapping within foundry to move the solidity code rather than noir code.
- Move the Noir project into a different directory.

### Additional context

_No response_

### Submission Checklist

- [X] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### LINKED ISSUE #2076: Nargo entry points should be configurable via Nargo.toml
### Problem

Currently, nargo has hardcoded logic to search for a `lib.nr` or `main.nr` file to determine a CrateType. Since the CrateType will be specified in Nargo.toml with #2073, we can decouple the paths and allow them to be configured via Nargo.toml with defaults for `lib.nr` or `main.nr` if not specified.

### Happy Case

A user can override the path to their entry point file, whether it is a lib, bin, or contract.

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

Yes

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by phated**
# Description

## Problem\*

Resolves #1040
Resolves #2076 

## Summary\*

This adds an optional `entry` field to the package section of a Nargo.toml which allows users to configure the entry point file that is processed by various commands. I also refactored the errors for missing entry files a little bit.

## Documentation

- [x] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [x] I will request for and support Dev Rel's help in documenting this PR.

  
  

When the Nargo.toml file is documented, we need to mention the optional `entry` field.

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/nargo_cli/tests/test_data/config.toml [hunk 1]
```diff
@@ -4,6 +4,7 @@ exclude = []
 # List of tests (as their directory name) expecting to fail: if the test pass, we report an error.
 fail = [
     "brillig_assert_fail",
+    "custom_entry_not_found",
     "dep_impl_primitive",
     "depend_on_bin",
     "workspace_fail",
```
Comment (TomAFrench) [inline crates/nargo_cli/tests/test_data/config.toml:10]: This breaks the regex in config.toml

**Review by TomAFrench — CHANGES_REQUESTED** [review]
Looks good. Main issue is in the config.toml

**Review by TomAFrench — APPROVED** [review]

**Comment by TomAFrench** [comment]
bumping this from the queue in favour of #2183

**Comment by phated** [comment]
> bumping this from the queue in favour of https://github.com/noir-lang/noir/pull/2183

There's a "Jump Queue" button on the MQ page for that

**Pull request merged**